File: web/src/lib/parsers/bradesco/csvExtrato.ts
// Parser do extrato da conta corrente do Bradesco em CSV. Mesma lógica de
// parsers/bradesco/csv_extrato.py (desktop) — ver ali para detalhes do
// formato observado (crédito/débito em colunas separadas, ruído filtrado
// pela tentativa de parse de data, marcador "COD. LANC." ignorado).
import type { BankParser, NormalizedRow } from '../types'
import { guessCategory, looksLikeInvestment } from '../base'
import { decodeBytes, parseBrlAmount, stripAccents } from '../common'
import { guessPaymentMethod } from './common'

function parseDateBr(s: string): string | null {
  const m = s.trim().match(/^(\d{2})\/(\d{2})\/(\d{4})$/)
  if (!m) return null
  return `${m[3]}-${m[2]}-${m[1]}`
}

export class BradescoCsvExtratoParser implements BankParser {
  bankId = 'bradesco'
  formatId = 'csv_extrato'

  sniff(bytes: ArrayBuffer, filename: string): boolean {
    if (!filename.toLowerCase().endsWith('.csv')) return false
    const text = decodeBytes(bytes.slice(0, 4096))
    const upper = stripAccents(text).toUpperCase()
    return upper.includes('HISTORICO') && upper.includes('DOCTO')
      && (upper.includes('CREDITO') || upper.includes('DEBITO'))
  }

  async parse(bytes: ArrayBuffer): Promise<NormalizedRow[]> {
    const text = decodeBytes(bytes)
    const lines = text.split(/\r?\n/).filter((ln) => ln.trim())

    const rows: NormalizedRow[] = []
    for (const ln of lines) {
      const parts = ln.split(';')
      if (parts.length < 6) continue
      const dateStr = parts[0].trim()
      const historico = parts[1].trim()
      const d = parseDateBr(dateStr)
      if (!d) continue
      if (stripAccents(historico).toUpperCase().startsWith('COD. LANC.')) continue

      const credito = parts[3].trim()
      const debito = parts[4].trim()
      let amount: number
      let direction: 'entrada' | 'saida'
      if (credito && credito !== '0,00') {
        amount = parseBrlAmount(credito)
        direction = 'entrada'
      } else if (debito && debito !== '0,00') {
        amount = parseBrlAmount(debito)
        direction = 'saida'
      } else {
        continue
      }

      const desc = historico
      const isInv = looksLikeInvestment(historico)
      rows.push({
        date: d,
        description: desc,
        amount: Math.abs(amount),
        direction,
        suggestedCategory: isInv ? 'Investimentos' : guessCategory(desc),
        suggestedPaymentMethod: guessPaymentMethod(historico),
        isInvestmentLike: isInv,
        isCreditCardCharge: false,
        raw: ln,
      })
    }
    return rows
  }
}

File: parsers/bradesco/csv_extrato.py
"""Parser do extrato da conta corrente do Bradesco em CSV.

Formato observado (amostra real, 2026-09): UTF-8 com BOM, separado por
";", cabeçalho "Data;Histórico;Docto.;Crédito (R$);Débito (R$);Saldo
(R$)" — ao contrário do Inter, crédito e débito são colunas separadas
(sem coluna única com sinal). O arquivo tem metadado antes do
cabeçalho, uma segunda mini-tabela "Últimos Lancamentos" com seu
próprio cabeçalho, e uma linha de Total no fim — em vez de procurar um
índice de cabeçalho, cada linha tenta virar uma data válida; o que não
vira é ruído e é ignorado de graça (mesmo truque que filtra tudo isso
sem precisar de lógica ciente de seção). Único caso à parte: linhas
"COD. LANC. N" são marcador de saldo (sem movimentação real), não
entram como transação.
"""
from datetime import datetime
from typing import List

from parsers.base import (
    NormalizedRow, decode_bytes, guess_category, looks_like_investment,
    parse_brl_amount, strip_accents,
)
from parsers.bradesco.common import guess_payment_method


class BradescoCsvExtratoParser:
    bank_id   = "bradesco"
    format_id = "csv_extrato"

    def sniff(self, data: bytes, filename: str) -> bool:
        if not filename.lower().endswith(".csv"):
            return False
        text = decode_bytes(data[:4096])
        upper = strip_accents(text).upper()
        return (
            "HISTORICO" in upper and "DOCTO" in upper
            and ("CREDITO" in upper or "DEBITO" in upper)
        )

    def parse(self, data: bytes) -> List[NormalizedRow]:
        text  = decode_bytes(data)
        lines = [ln for ln in text.splitlines() if ln.strip()]

        rows: List[NormalizedRow] = []
        for ln in lines:
            parts = ln.split(";")
            if len(parts) < 6:
                continue
            date_str  = parts[0].strip()
            historico = parts[1].strip()
            try:
                d = datetime.strptime(date_str, "%d/%m/%Y").date()
            except ValueError:
                continue
            if strip_accents(historico).upper().startswith("COD. LANC."):
                continue

            credito = parts[3].strip()
            debito  = parts[4].strip()
            if credito and credito != "0,00":
                amount, direction = parse_brl_amount(credito), "entrada"
            elif debito and debito != "0,00":
                amount, direction = parse_brl_amount(debito), "saida"
            else:
                continue

            desc = historico
            is_inv = looks_like_investment(historico)
            rows.append(NormalizedRow(
                date=d,
                description=desc,
                amount=abs(amount),
                direction=direction,
                suggested_category="Investimentos" if is_inv else guess_category(desc),
                suggested_payment_method=guess_payment_method(historico),
                is_investment_like=is_inv,
                raw=ln,
            ))
        return rows
